RocketChat/Rocket.Chat
# [FIX] Message search breaking URL and emojis, and missing results [WIP]

## Proposed changes

It has been reported that when searching for a message, the results were displaying with bugs : URLs where broken, emojis title too, and some results are missing. This PR aims to fix these issues.

Elements resolved:
- [X] URLs are broken by search `<mark>`
- [X] Emojis are broken by search `<mark>`
- [X] `mailto:` links  are broken by search `<mark>`
- [ ] Some search results are missing (is it really a bug ? See comment below)

## Issue(s)

#18770 => reports that highlighted links, emojis (and mentions ?) are broken, as well as missing results
#18696 => reports that highlighted links are broken
#18456 => report that highlighted links are broken

## How to test or reproduce

Go into a channel, and post these messages :
> smile
> :smile:
> smile.com
> smile-not.com
> smilenot.com

Then search for the word `smile` in that channel. You'll notice that: 
- you can't click on links anymore (they're broken by `<mark>` tags)
- if you hover on the smile emoji, its title is broken by `<mark>` tags too
- the `smilenot.com` link isn't part of the results

## Screenshots
See issues for screenshots

## Types of changes




## Checklist



## Changelog




## Further comments


#### Broken links/emojis
I fixed the `<mark>` tag issue by doing three `replace` steps
1. Using a (bit tricky) REGEX, we find every occurence of `searchedText` which is inside a `title="..."` or `href="..."` and replace `searchedText` in them by an unique UUID
2. Using the previous line of code, we find every occurence of `searchedText` left and we mark it
3. We find every occurence of the UUID and replace them with `searchedText`

I did this because I couldn't make a better REGEX, which would exclude `1` and find only `2`. If you are able to do it, or if you see a better solution, feel free to comment ! Also, you may wonder "Why an UUID ?" : I used it because I thought that any arbitrary set of characters could end up being part of the initial search, and I think we're safe about that with a V4 UUID.

#### Missing results
Considering the other issue, which we'll call "missing results", I searched in the whole project and found that it uses a **Meteor function `messageSearch`**. So I suspect Meteor for being responsible of that bug, but we need more investigation to be sure.

#### Related PR
Results highlighting has been implemented in #16166
